### PR TITLE
New architecture

### PR DESCRIPTION
**Pull request by EclipseGc**
Migrate Service definitions to leverage contexts, enable serialization, solve page cache problems, respect accept headers, clear cache when the context objects (like a node) are update and a few other things I've probably forgotten.

### CODE REVIEW

File: src/Controller/Services.php [hunk 1]
```diff
@@ -18,18 +23,72 @@
...
+  /**
+   * @var \Symfony\Component\Serializer\SerializerInterface
    */
+  protected $serializer;
 
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container) {
+    return new static($container->get('plugin.manager.services.service_definition'), $container->get('serializer'));
```
Comment (nerdstein) [inline src/Controller/Services.php]: Can you help explain how the container knows the serializer? Just for my knowledge

Comment (EclipseGc) [inline src/Controller/Services.php]: This is a class that's knowledgable about ALL serializers. We pass the format later and it serializes to a specific type at that point.

File: src/Controller/Services.php [hunk 2]
```diff
@@ -18,18 +23,72 @@
...
+     * Iterate over any contexts defined for this plugin and extract them from
+     * the request defaults if the naming is identical. This means that a
+     * context named 'node' would match to a url parameter {node} or a route
+     * default named 'node'.
+     */
+    foreach ($service_def->getContextDefinitions() as $context_id => $context_definition) {
+      if ($request->attributes->has($context_id)) {
+        $context = new Context($context_definition);
+        $context->setContextValue($request->attributes->get($context_id));
+        $service_def->setContext($context_id, $context);
```
Comment (nerdstein) [inline src/Controller/Services.php]: This design is great

File: src/Controller/Services.php (hunk 1 above)
Comment (nerdstein) [inline src/Controller/Services.php]: Ah ok, awesome

File: src/ServiceDefinitionInterface.php [hunk 3]
```diff
@@ -51,12 +51,12 @@ public function supportsTranslation();
...
-   *   A request object.
-   * @param route_match
-   *   A route match object.
+   * @param \Symfony\Component\HttpFoundation\Request $request
+   *   The request object.
+   * @param \Drupal\Core\Routing\RouteMatchInterface $route_match
+   *   The route match object.
    *
-   * @return SerializationInterface
+   * @return array
```
Comment (nerdstein) [inline src/ServiceDefinitionInterface.php]: We should document this in the README

**Pull request merged**